File: TypeScript/src/algorithms/sorting/mergeSort.ts
const merge = (arrayA: number[], start: number, middle: number, end: number) => {
  let i = start,
    j = middle,
    k = 0

  const sortedArray = arrayA.slice()
  const arrayB: number[] = []

  while (i < middle && j < end) {
    if (sortedArray[i] <= sortedArray[j]) {
      arrayB[k++] = sortedArray[i++]
    } else {
      arrayB[k++] = sortedArray[j++]
    }
  }

  while (i < middle) {
    arrayB[k++] = sortedArray[i++]
  }

  while (j < end) {
    arrayB[k++] = sortedArray[j++]
  }

  for (let i = start; i < end; i++) {
    sortedArray[i] = arrayB[i - start]
  }

  return sortedArray
}

export const mergeSort = (array: number[], start: number, end: number) => {
  let sortedArray: number[] = []
  if (start < end - 1) {
    let middle = (start + end) / 2

    mergeSort(array, start, middle)
    mergeSort(array, middle, end)
    sortedArray = merge(array, start, middle, end)
  }

  return sortedArray
}

File: Go/algorithms/sorting/mergeSort/mergeSort.go
package sorting

func merge(arr []int, start, middle, end int){
	i, j, k := start, middle, 0
	slice := make([]int, end-start)
	for i < middle && j < end {
		if arr[i] <= arr[j] {
			slice[k] = arr[i]
			k++
			i++
		} else {
			slice[k] = arr[j]
			k++
			j++
		}
	}
	for i < middle {
		slice[k] = arr[i]
		k++
		i++
	}
	for j < end {
		slice[k] = arr[j]
		k++
		j++
	}
	for i := start; i < end; i++ {
        arr[i] = slice[i - start]
    }
}

func MergeSort(arr []int, start, end int) {
    if start < end - 1 {
        middle := (start + end) / 2;
        MergeSort(arr, start, middle)
        MergeSort(arr, middle, end)
        merge(arr, start, middle, end)
    }
}
